src/Kestrel.Core/Internal/ConnectionLimitMiddleware.cs

```diff
@@ -1,32 +1,74 @@
-﻿using System.Threading.Tasks;
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Threading.Tasks;
 using Microsoft.AspNetCore.Protocols;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Features;
 using Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure;
 
 namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal
 {
     public class ConnectionLimitMiddleware
     {
-        private readonly ServiceContext _serviceContext;
         private readonly ConnectionDelegate _next;
+        private readonly ResourceCounter _concurrentConnectionCounter;
+        private readonly IKestrelTrace _trace;
 
-        public ConnectionLimitMiddleware(ConnectionDelegate next, ServiceContext serviceContext)
+        public ConnectionLimitMiddleware(ConnectionDelegate next, long connectionLimit, IKestrelTrace trace)
+            : this(next, ResourceCounter.Quota(connectionLimit), trace)
+        {
+        }
+
+        // For Testing
+        internal ConnectionLimitMiddleware(ConnectionDelegate next, ResourceCounter concurrentConnectionCounter, IKestrelTrace trace)
         {
             _next = next;
-            _serviceContext = serviceContext;
+            _concurrentConnectionCounter = concurrentConnectionCounter;
+            _trace = trace;
         }
 
-        public Task OnConnectionAsync(ConnectionContext connection)
+        public async Task OnConnectionAsync(ConnectionContext connection)
         {
-            if (!_serviceContext.ConnectionManager.NormalConnectionCount.TryLockOne())
+            if (!_concurrentConnectionCounter.TryLockOne())
             {
                 KestrelEventSource.Log.ConnectionRejected(connection.ConnectionId);
-                _serviceContext.Log.ConnectionRejected(connection.ConnectionId);
+                _trace?.ConnectionRejected(connection.ConnectionId);
```

Comment:
trace can be null?

Reply:
No unless someone constructs it manually. It could have been before when there was the IConnectionBuilder extension method.